fvm/transactionStorageLimiter.go

```diff
@@ -14,29 +14,76 @@ import (
 
 type TransactionStorageLimiter struct{}
 
-func (_ TransactionStorageLimiter) CheckStorageLimits(
+// CheckStorageLimits checks each account that had its storage written to during a transaction, that its storage used
+// is less than its storage capacity.
+// Storage used is an FVM register and is easily accessible.
+// Storage capacity is calculated by the FlowStorageFees contract from the account's flow balance.
+//
+// The payers balance is considered to be maxTxFees lower that its actual balance, due to the fact that
+// the fee deduction step happens after the storage limit check.
+func (d TransactionStorageLimiter) CheckStorageLimits(
 	env environment.Environment,
 	addresses []flow.Address,
+	payer flow.Address,
+	maxTxFees uint64,
 ) error {
 	if !env.LimitAccountStorage() {
 		return nil
 	}
 
 	defer env.StartSpanFromRoot(trace.FVMTransactionStorageUsedCheck).End()
 
+	err := d.checkStorageLimits(env, addresses, payer, maxTxFees)
+	if err != nil {
+		return fmt.Errorf("storage limit check failed: %w", err)
+	}
+	return nil
+}
+
+func (d TransactionStorageLimiter) checkStorageLimits(
+	env environment.Environment,
+	addresses []flow.Address,
+	payer flow.Address,
+	maxTxFees uint64,
+) error {
 	commonAddresses := make([]common.Address, len(addresses))
 	usages := make([]uint64, len(commonAddresses))
+
+	commonAddressesContainPayer := false
+
```

Comment:
nit: maybe first pre-process addresses so that we can remove the code duplication below?
(this would also avoid potential reallocation of `commonAddresses`/`usages` in cases where they are off by one)
```go
if maxTxFees > 0 {
	commonAddressesContainPayer := false
	for _, address := range addresses {
		if address == payer {
			commonAddressesContainPayer = true
			break
		}
	}

	if !commonAddressesContainPayer {
		addresses = append(addresses, payer)
	}
}
```